### DISCUSSION TITLE
Can field indexer be simplify?

### DISCUSSION THREAD

**Discussion by linonetwo** [TiddlyWiki/TiddlyWiki5#7278] [Q&A]
https://github.com/Jermolene/TiddlyWiki5/blob/95dc56d8508109a76c232cfbcd858889d32916a9/core/modules/indexers/field-indexer.js#L35-L41

Can it be

```
	this.wiki.each.byField = function(name,value) {
		var lookup = self.lookup(name,value);
		return lookup;
	};
```

Why check against the whole wiki? This is the most time consuming part. I'm not sure I understand this, so this is a discussion.

![截屏2023-02-22 14 08 27](https://user-images.githubusercontent.com/3746270/220537289-ba6f2927-6ba1-4458-af64-aa26f3980373.png)

The `lookup` here is already a list of tiddler titles.

If you want to update the indexer by removing titles that is not longer exist, we can do this in an deferred async task.

**Comment by pmario** [comment]
Did you run the tests with your changes applied?

**Reply by linonetwo** [reply]
I have tried, and It works without any obvious bug, and the speed is dramatically increased.

**Reply by linonetwo** [reply]
Oh, you mean tw test set, I'm trying.

**Reply by linonetwo** [reply]
Randomized with seed 29235
Started
.......................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................


775 specs, 0 failures
Finished in 3.369 seconds
Randomized with seed 29235 (jasmine --random=true --seed=29235)
To run the tests in a browser, open editions/test/output/test.html

**Comment by Jermolene** [comment]
Hi @linonetwo I think that the issue is that `self.lookup(name,value);` includes shadow tiddlers in the index, and `wiki.each.byField()` is intended to only return non-shadow tiddlers.

**Reply by linonetwo** [reply]
But `titles.indexOf` make it O(n^2) , can we use simplier approach here, e.g. `$tw.wiki.isShadowTiddler` which use `hasOwnProperty` is O(1).

Or we can just index non-shadow, just replace `this.wiki.eachTiddlerPlusShadows` with `this.wiki.forEachTiddler` as long as we don't need shadow.

If we need shadow, we can even create `this.index` and `this.shadowIndex`, exchange time with space.

**Reply by linonetwo** [reply]
I see, we have `wiki.eachShadowPlusTiddlers.byField`. So we also need to index shadow tiddlers.

Let me create `this.index` and `this.shadowIndex`? Because `wiki.each.byField` used more frequently by plugin, we need to make it efficient.

**Reply by linonetwo** [reply]
Oh, that is too complicated, use `$tw.wiki.isShadowTiddler` instead.

**Reply by pmario** [reply]
>But titles.indexOf make it O(n^2) 

IMO titles.indexOf is O(n) ... titles is an array. .. but anyway O(1) is preferable.

**Reply by linonetwo** [reply]
By n^2 I mean it was `.filter`'s n x `.indexOf`'s n. This is the case we might optimize for in the future.

If we have a situation to do so, build an object outside of `.filter` as hash map will make it faster too. Thankfully in this case we already have a hash map to use (used by `isShadowTiddler`).
